### DISCUSSION TITLE
posts/Go/singleton_in_go/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#343] [Announcements]
https://www.liwenzhou.com/posts/Go/singleton_in_go/ 

 李文周的Blog Go语言教程 golang中文教程 李文周博客 七米 qimi Go语言并发 并发元语 goroutine sync 单例模式 并发安全 singleton 设计模式

**Comment by evanzme** [comment]
单例模式使用
once.Do(func() {
    instance = &singleton{}
})
如果在并发场景下，第一个初始化失败了，后面就再也不会执行了，这个问题是不是也需要处理一下？

**Comment by wjhtime** [comment]
if initialized == 0 {
         instance = &singleton{}
         atomic.StoreUint32(&initialized, 1)
    }
为什么不是 if atomic.LoadUInt32(&initialized) == 0

**Reply by asyncpjax** [reply]
锁的内存屏障作用
`mu.Lock()` 和 `mu.Unlock()` 内部实现了 全屏障（Full Memory Barrier）：
锁获取（Lock）：保证后续代码的内存读写不会重排到锁获取之前。
锁释放（Unlock）：保证之前代码的内存读写不会重排到锁释放之后。
实际效果：在 `mu.Lock()` 后，所有 goroutine 对共享变量的修改已同步到当前线程的内存视图中。
结论：在锁的保护下，直接读取 `initialized == 0` 是安全的，无需原子操作。
AI回复
